Source/WTF/wtf/threads/Signals.cpp

```diff
@@ -203,6 +206,157 @@ kern_return_t catch_mach_exception_raise_state_identity(mach_port_t, mach_port_t
     return KERN_FAILURE;
 }
 
+#ifdef EXCEPTION_IDENTITY_PROTECTED
+
+static thread_act_t threadIDGetThreadPort(uint64_t threadID)
+{
+    kern_return_t kr;
+    mach_msg_type_number_t i, thcnt;
+    thread_act_array_t threads;
+    struct thread_identifier_info tiinfo;
+    unsigned infoCount = THREAD_IDENTIFIER_INFO_COUNT;
+    thread_act_t thread = THREAD_NULL;
+
+    task_threads(mach_task_self(), &threads, &thcnt);
+    for (i = 0; i < thcnt; i++) {
```

Comment:
Why do we have to iterate through every thread in the process? That seems like it shouldn't be necessary...

Reply:
I agree it seems a bit ridiculous, but unfortunately this is the only way that the kernel supports for matching threadID to threadPort. do you think it will be a significant problem?

Reply:
It's hypothetically performance sensitive as this code can be triggered by WebAssembly out of bounds memory access (we reserve a 4GB virtual address space for WebAssembly and only map RW live pages so the CPU handles bounds checking for us). 

It seems like EXCEPTION_STATE_PROTECTED would give the port, which avoids this dance. What's the advantage from a security perspective of EXCEPTION_IDENTITY_PROTECTED?

Reply:
The issue security-wise is in the giving of the port directly. There have been previous vulnerabilities where a malicious process can steal the exception port of a privileged process, and then if you trigger an exception the thread/task ports are now sent over to the malicious process (which is very bad). By forcing everyone to use these tokens, the kernel can now verify that when it hands out a thread/task port in exchange for one of these tokens, it is not handing these out to a malicious process (something that is impossible to do at the time of sending the exception).